core/solver/cb_gmres.cpp

```diff
@@ -493,15 +505,16 @@ void CbGmres<ValueType>::apply_impl(const LinOp *b, LinOp *x) const
 
 template <typename ValueType>
 void CbGmres<ValueType>::apply_impl(const LinOp *alpha, const LinOp *b,
-                                    const LinOp *residual_norm_collection,
-                                    LinOp *x) const
+                                    const LinOp *beta, LinOp *x) const
 {
-    auto dense_x = as<matrix::Dense<ValueType>>(x);
-
-    auto x_clone = dense_x->clone();
-    this->apply(b, x_clone.get());
-    dense_x->scale(residual_norm_collection);
-    dense_x->add_scaled(alpha, x_clone.get());
+    precision_dispatch_real_complex<ValueType>(
+        [this](auto dense_alpha, auto dense_b, auto dense_beta, auto dense_x) {
+            auto x_clone = dense_x->clone();
+            this->apply(dense_b, x_clone.get());
```

Comment:
Some minor improvement? Now you could directly call `apply_dense_impl`. Same everywhere.
```suggestion
            this->apply_dense_impl(dense_b, x_clone.get());
```

Reply:
Good idea, thanks!